### PR TITLE
Firestore Data Linked to Geochart

### PR DESCRIPTION
**Pull request by kirakira0**
image

If possible, I would especially appreciate optimisation-related feedback. The geochart loads very slowly.
(Edit 2020/07/23: I have significantly improved load times using Promise.all() as opposed to a sequence of awaits.)

NOTE: I see that Github is showing FirebaseAdReader as a new file. I'm not sure why that is. This same file has appeared in previous PRs and I did not rename it or delete it and create a new file with the same name.

NOTE: StateSubcollectionsBuilder serves the same purpose as a file that appears in previous PRs called "StateCollectionsBuilder". It underwent pretty heavy refactoring and a name change so I created a new file for it because I was confusing myself. I'm not sure if this is bad practice. If it is, as I suspect it might be, please let me know and I will refrain from doing it again in the future.

### CODE REVIEW

File: client/src/constants/geochart_constants.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Defines set of states. 
+ * Used for Geochart queries.
+ * @returns {Set} state set
```
Comment (rob-marcus) [inline client/src/constants/geochart_constants.js:4]: very minor nit but this doesn't return, it's just a `const`.

File: client/src/components/Visualization/charts/Geochart.js [hunk 2]
```diff
@@ -4,45 +4,74 @@
...
-    for (const state in states) {
-      data.push([state, Math.floor(Math.random() * Math.floor(1000))]);
+
+  const [adTotal, setAdTotal] = useState(0);
+  const WIDTH = "700";
+  const HEIGHT = "400";
+
+  useEffect(() => {
+    async function fetchStateTotals() {
+
```
Comment (rob-marcus) [inline client/src/components/Visualization/charts/Geochart.js:20]: Nit: unnecessary whitespace.

File: client/src/components/Visualization/charts/Geochart.js [hunk 3]
```diff
@@ -4,45 +4,74 @@
...
+      let data = [["State", "Total Ad Spend (USD)", {type: 'string', role: 'tooltip'}]];
```
Comment (rob-marcus) [inline client/src/components/Visualization/charts/Geochart.js:21]: This is kind of an odd object, a comment would be nice.

File: client/src/components/Visualization/charts/Geochart.js [hunk 4]
```diff
@@ -4,45 +4,74 @@
...
+
+      for (let state of STATES) {
+
+        // Calculate state spend totals.
+        let stateTotal = 0;
+        let stateSpendData = await database.collection('dev_states')
```
Comment (rob-marcus) [inline client/src/components/Visualization/charts/Geochart.js:27]: Since you use `dev_states` in multiple places, and since this is being pushed to master, i.e., not `dev`, please change `dev_states` to `prod_states` or whatever and declare a const parameter like `states_collection = database.collection(states_collection_name)`.

File: client/src/components/Visualization/charts/Geochart.js [hunk 5]
```diff
@@ -4,45 +4,74 @@
...
+                                           .doc(state.toLowerCase())
+                                           .collection('stateData')
+                                           .doc('spendData')
+                                           .get();
+        // Only add to total if present.
+        if (stateSpendData.data()) {
```
Comment (rob-marcus) [inline client/src/components/Visualization/charts/Geochart.js:33]: This is a tad bit confusing because you are using `stateSpendData.data()` in a truthy way, but it probably isn't a boolean. Something better might be `stateSpendData.data() !== undefined`.

File: client/src/components/Visualization/charts/Geochart.js [hunk 6]
```diff
@@ -4,45 +4,74 @@
...
+
+      for (let state of STATES) {
+
+        // Calculate state spend totals.
+        let stateTotal = 0;
+        let stateSpendData = await database.collection('dev_states')
+                                           .doc(state.toLowerCase())
+                                           .collection('stateData')
+                                           .doc('spendData')
+                                           .get();
```
Comment (rob-marcus) [inline client/src/components/Visualization/charts/Geochart.js:31]: How big is this query? Reason I'm asking is bc it seems like you only want one field, in which case, why not just have a separate document for that field so you don't have to wait for `n` things to load?

File: client/src/components/Visualization/charts/Geochart.js [hunk 7]
```diff
@@ -4,45 +4,74 @@
...
+        // Calculate top advertiser per state.
+        let topAdvertiser = '';
+        let advertiserCollectionRef = database.collection('dev_states')
+                                              .doc(state.toLowerCase())
+                                              .collection('advertisers')
+                                              .orderBy('totalAdvertiserSpend')
+                                              .limit(1);
+        let topAdvertiserCollection = await advertiserCollectionRef.get();
+        // Only add to total if present.
+        if (topAdvertiserCollection.docs[0]) {
```
Comment (rob-marcus) [inline client/src/components/Visualization/charts/Geochart.js:46]: Same thing about using this in a `truthy` way. Confusing.

File: client/src/components/Visualization/charts/Geochart.js [hunk 8]
```diff
@@ -4,45 +4,74 @@
...
-        data={getData()}
-        options={options}
-      />
+      <p>Ad Spend Geochart</p>
+      { adTotal.length > 0 ? <Chart chartType="GeoChart" 
+                                    width={WIDTH} 
+                                    height={HEIGHT} 
+                                    data={adTotal} 
+                                    options={options} /> 
+      : <p>Loading Ad Spend Geochart...</p> }
```
Comment (rob-marcus) [inline client/src/components/Visualization/charts/Geochart.js:74]: Indent this +4 spaces so it's clear this is still part of the ternary operation.

File: server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+import com.google.cloud.firestore.WriteResult;
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.FirebaseOptions;
+import com.google.firebase.cloud.FirestoreClient;
+import com.google.sps.utils.Ad;
+import com.google.sps.utils.Constants;
+import java.io.FileInputStream;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
```
Comment (rob-marcus) [inline server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:22]: Importing unused dependencies. InputStream, array list, arrays, hashset.

File: server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+      String advertiser = document.getString("advertiser");
+      long spendMax = document.getLong("spendMax");
+      if (advertiserToTotalSpend.containsKey(advertiser)) {
+        advertiserToTotalSpend.put(advertiser, 
+          advertiserToTotalSpend.get(advertiser) + spendMax);
+      } else {
+        advertiserToTotalSpend.put(advertiser, spendMax);
+      }
+    }
+    return advertiserToTotalSpend;
```
Comment (rob-marcus) [inline server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:80]: Minor spacing nit: this is a pretty dense function/block of code (insert [chunky Putin](https://www.youtube.com/watch?v=nDanLjEnz7Q)). 

An easy way to improve readability would be to like  move line 69 to 68, ad whitespace between 'long' and 'if' and add whitespace between `return` and `}`.

File: server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+    long totalStateSpend = 0;
+
+    for (String key : advertiserToTotalSpend.keySet()) {
+      Map<String, Object> data = new HashMap<>();
+      long totalAdvertiserSpend = advertiserToTotalSpend.get(key);
+      totalStateSpend += totalAdvertiserSpend;
+      data.put("totalAdvertiserSpend", totalAdvertiserSpend);
+      db
+        .collection(WRITE_COLLECTION).document(state.toLowerCase())
```
Comment (rob-marcus) [inline server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:101]: rogue wrapping!

File: server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+    }
+    return advertiserToTotalSpend;
+  }
+
+  /**
+   * Writes advertiser documents to Firestore
+   * @param state the current state subcollection (tells Firestore
+   *              where to write the advertiser documents)
+   * @param advertiserToTotalSpend maps that relates every advertiser
+   *              to their total ad spend in a particular stateit bran
```
Comment (rob-marcus) [inline server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:88]: `stateit bran`?

File: server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+    for (String key : advertiserToTotalSpend.keySet()) {
+      Map<String, Object> data = new HashMap<>();
+      long totalAdvertiserSpend = advertiserToTotalSpend.get(key);
+      totalStateSpend += totalAdvertiserSpend;
+      data.put("totalAdvertiserSpend", totalAdvertiserSpend);
+      db
+        .collection(WRITE_COLLECTION).document(state.toLowerCase())
+        .collection("advertisers").document(key)
+        .set(data, SetOptions.merge());
```
Comment (rob-marcus) [inline server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:103]: I had to look up what `SetOptions.merge()` did. Neat! Would recommend a comment though that you aren't overwriting, just adding a field (I think?)

File: server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+    db = FirestoreClient.getFirestore();
+    return db;
+  }
+
+  /**
+   * Maps advertisers to their total ad spend in a particular state
+   * @param documentsInState list of all document snapshots targeted
+   *                         at a particular state 
+   */
+  public static Map<String, Long> getAdvertiserToTotalSpend(
```
Comment (rob-marcus) [inline server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:66]: borderline awkward name. `ToTotal`. Consider dropping the `To`?

Comment (kirakira0) [reply server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:66]: I can remove the "spend". Will it still be clear that it's a dictionary if I drop the "to"?

Comment (rob-marcus) [reply server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:66]: It's implicit from the return type that it's a dictionary.

Comment (kirakira0) [reply server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:66]: True although I don't want it to be read as "advertiser spend"

Comment (rob-marcus) [reply server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:66]: fwiw I think `advertiserSpending` or `mapAdvertiserSpending` is probably ok. I'm not the best namer so I'll defer to you/others.

File: client/src/components/Visualization/charts/Geochart.js (hunk 6 above)
Comment (kirakira0) [reply client/src/components/Visualization/charts/Geochart.js:33]: That was a good suggestion, thank you. I've moved totalStateSpend up so that it's now a field of the top level state document & refactored the query accordingly. The change shows up in the "Files Changed" but not in this comment, so I'm resolving to show that it's been addressed.

File: server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+import com.google.cloud.firestore.WriteResult;
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.FirebaseOptions;
+import com.google.firebase.cloud.FirestoreClient;
+import com.google.sps.utils.Ad;
+import com.google.sps.utils.Constants;
+import java.io.FileInputStream;
+// import java.io.InputStream;
+// import java.util.ArrayList;
+// import java.util.Arrays;
```
Comment (ailemeo) [inline server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:22]: Delete?

File: server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+    for (String key : advertiserToSpend.keySet()) {
+      Map<String, Object> data = new HashMap<>();
+      long totalAdvertiserSpend = advertiserToSpend.get(key);
+      totalStateSpend += totalAdvertiserSpend;
+      data.put("totalAdvertiserSpend", totalAdvertiserSpend);
+      db.collection(WRITE_COLLECTION)
+        .document(state.toLowerCase())
+        .collection("advertisers")
+        .document(key)
+        .set(data, SetOptions.merge()); // Granular merge instead of overwrite.
```
Comment (ailemeo) [inline server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:106]: Does Firestore have some kind of bulk update API, so you don't have to do this one-by-one in a loop?

File: server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+    db = FirestoreClient.getFirestore();
+    return db;
+  }
+
+  /**
+   * Maps advertisers to their total ad spend in a particular state
+   * @param documentsInState list of all document snapshots targeted
+   *                         at a particular state 
+   */
+    public static Map<String, Long> getAdvertiserToSpend(
```
Comment (ailemeo) [inline server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:67]: Rename to `getAdvertiserSpend` 
The current name sounds like you're getting an advertiser that you intend to "spend"

File: server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+      List<QueryDocumentSnapshot> documentsInState) {
+    Map<String, Long> advertiserToSpend = new HashMap<>();
```
Comment (ailemeo) [inline server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:69]: Nit: advertiserToSpendMap

File: server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+
+    for (QueryDocumentSnapshot document: documentsInState) {
+      String advertiser = document.getString("advertiser");
+      long spendMax = document.getLong("spendMax");
```
Comment (ailemeo) [inline server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:73]: Nit: maxSpend/maximumSpend

File: server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+  }
+
+  /**
+   * Writes advertiser documents to Firestore
+   * @param state the current state subcollection (tells Firestore
+   *              where to write the advertiser documents)
+   * @param advertiserToSpend map that relates every advertiser
+   *              to their total ad spend in a particular state              
+   */
+  public static void updateStateSubcollection(String state,
```
Comment (ailemeo) [inline server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:92]: Are there tests for these methods?

File: server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+ * Author: Kira Toal
+ * Date: July 20, 2020
+ */ 
+public class StateSubcollectionBuilder {
+
+  private static final String PATH_TO_SERVICE_ACCOUNT = "./serviceAccountKey.json"; 
+  private static final String DATABASE_URL = "https://step9-2020-capstone.firebaseio.com"; 
+  private static final String MAIN_COLLECTION = "ads";
+  private static final String WRITE_COLLECTION = "dev_states";
+  private static final Set<String> GEO_TARGETS = Constants.TEST_GEO_TARGETS;
```
Comment (ailemeo) [inline server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:43]: Didn't you just delete TEST_GEO_TARGETS?

File: client/src/components/Visualization/charts/Geochart.js [hunk 22]
```diff
@@ -1,76 +1,80 @@
...
+        the table ("Total Ad Spend (USD)").
+      The first column (state name) and third column (info on total state ad 
+        spend and top advertiser) are shown in the tooltip.
+      */ 
+      let dataTable = [["State", "Total Ad Spend (USD)", {type: 'string', role: 'tooltip'}]];
+      for (let state of STATES) {
+        
+        // Calculate state spend totals.
+        let stateTotal = 0;
+        let stateSpendData = await database.collection(STATES_COLLECTION)
```
Comment (ailemeo) [inline client/src/components/Visualization/charts/Geochart.js:31]: Why are the database.collection calls here being done sequentially with `await`s, especially since they're independent. Might be contributing to the long latency.

Comment (ailemeo) [reply client/src/components/Visualization/charts/Geochart.js:31]: Also general debugging advice: you should be able to use the "Networks" tab in chrome dev tools to track the duration of the server calls

File: server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java (hunk 20 above)
Comment (kirakira0) [reply server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:92]: Not in this PR, but I'm currently working on one with tests and cloud functions to live update the states subcollections

File: client/src/components/Visualization/charts/Geochart.js [hunk 23]
```diff
@@ -1,59 +1,67 @@
...
+            .collection('advertisers')
+            .orderBy('totalAdvertiserSpend')
+            .limit(1)
+            .get()
+        );
+      }
+
+      let stateResults = await Promise.all(stateSpendPromises);
+      let advertiserResults = await Promise.all(topAdvertiserPromises);
+      const zippedResults = stateResults.map((e, i) => [e,advertiserResults[i]]);
```
Comment (ailemeo) [inline client/src/components/Visualization/charts/Geochart.js:47]: Nit: add space
`[e, advertiserResults[i]]`

File: client/src/components/Visualization/charts/Geochart.js [hunk 24]
```diff
@@ -1,59 +1,67 @@
...
 
-  useEffect(async () => {
-    const data = [['State', 'Total Ad Spend Across All Advertisers']];
-    for (const state in states) {
-      data.push([
-        state,
-        Math.floor(Math.random() * Math.floor(250000) + 10000),
-      ]); // Update data table.
+      for (let result of zippedResults) {
```
Comment (ailemeo) [inline client/src/components/Visualization/charts/Geochart.js:49]: You can make use of ES6 destructuring here.

`for (let [stateTotal, topAdvertiser] of zippedResults) {`

File: client/src/components/Visualization/charts/Geochart.js [hunk 25]
```diff
@@ -1,59 +1,67 @@
...
-import { Chart } from 'react-google-charts';
-import { states } from './StateDataParser';
+import { Chart } from "react-google-charts";
+import React, { useState, useEffect } from 'react';
+import { STATES } from '../../../constants/geochart_constants';
+import { database } from '../../../firebase/firebase';
 
 const Geochart = () => {
+  const STATES_COLLECTION = 'dev_states';
   const [adTotal, setAdTotal] = useState(0);
```
Comment (ailemeo) [inline client/src/components/Visualization/charts/Geochart.js:13]: How does this work? 
I assume the second element is used to send data to the chart since you call it with your `dataTable` below.

What's the first element for?

File: server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java [hunk 26]
```diff
@@ -0,0 +1,135 @@
...
+    
+    // Set state spend.
+    DocumentReference stateRef = db.collection(WRITE_COLLECTION)
+        .document(state);
+    Map<String, Object> stateSpendMap = new HashMap<>();
+    stateSpendMap.put("totalStateSpend", totalStateSpend);
+    batch.set(stateRef, stateSpendMap);
+
+    // Commit the batch.
+    batch.commit();
```
Comment (ailemeo) [inline server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:119]: Looks much better.

File: client/src/components/Visualization/charts/Geochart.js (hunk 25 above)
Comment (ailemeo) [reply client/src/components/Visualization/charts/Geochart.js:13]: Ignore this. I understand now from looking at a different PR.

File: client/src/components/Visualization/charts/Geochart.js [hunk 27]
```diff
@@ -1,59 +1,67 @@
...
-    const data = [['State', 'Total Ad Spend Across All Advertisers']];
-    for (const state in states) {
-      data.push([
-        state,
-        Math.floor(Math.random() * Math.floor(250000) + 10000),
-      ]); // Update data table.
+      for (let result of zippedResults) {
+        let stateTotal = result[0];
+        let topAdvertiser = result[1];
+        if (stateTotal.data() !== undefined) {
```
Comment (andrewlathamgoogle) [inline client/src/components/Visualization/charts/Geochart.js:52]: Under what circumstances would data be undefined?  Should you throw an error?

File: client/src/constants/geochart_constants.js [hunk 28]
```diff
@@ -0,0 +1,15 @@
...
+ */
+const STATES = new Set(["Alabama", "Alaska", "Arizona", "Arkansas", "California",
+    "Colorado", "Connecticut", "Delaware", "Florida", "Georgia", "Hawaii",
+    "Idaho", "Illinois", "Indiana", "Iowa", "Kansas", "Kentucky", "Louisiana",
+    "Maine", "Maryland", "Massachusetts", "Michigan", "Minnesota", "Mississippi",
+    "Missouri", "Montana", "Nebraska", "Nevada", "New Hampshire", "New Jersey",
+    "New Mexico", "New York", "North Carolina", "North Dakota", "Ohio", "Oklahoma",
+    "Oregon", "Pennsylvania", "Rhode Island", "South Carolina", "South Dakota",
+    "Tennessee", "Texas", "Utah", "Vermont", "Virginia", "Washington", 
+    "West Virginia", "Wisconsin", "Wyoming", "United States"]);
```
Comment (andrewlathamgoogle) [inline client/src/constants/geochart_constants.js:13]: The United States isn't a state.

File: server/src/main/java/com/google/sps/utils/ReportReader.java [hunk 29]
```diff
@@ -25,8 +25,8 @@
   // NUMBER_CSV_HEADER_ROWS helps the parser know what rows to skip when reading the CSV.
   private static final int NUMBER_CSV_HEADER_ROWS = 2;
   private static final String COLLECTION = "ads";
-  private static final int START_ROW_INDEX = 1;  // First row sent to processor (inclusive).
-  private static final int END_ROW_INDEX = 5000; // Last row sent to processor (inclusive).
+  private static final int START_ROW_INDEX = 20001;  // First row sent to processor (inclusive).
```
Comment (andrewlathamgoogle) [inline server/src/main/java/com/google/sps/utils/ReportReader.java:28]: ?

File: server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java [hunk 30]
```diff
@@ -0,0 +1,135 @@
...
+  /**
+   * Writes advertiser documents to Firestore
+   * @param state the current state subcollection (tells Firestore
+   *              where to write the advertiser documents)
+   * @param advertiserSpendMap map that relates every advertiser
+   *              to their total ad spend in a particular state              
+   */
+  public static void updateAdvertiserSpendSubCollection(String state,
+      Map<String, Long> advertiserSpendMap) {
+
```
Comment (andrewlathamgoogle) [inline server/src/main/java/com/google/sps/utils/StateSubcollectionBuilder.java:93]: nit:  unnecessary spacing

File: client/src/constants/geochart_constants.js (hunk 28 above)
Comment (kirakira0) [reply client/src/constants/geochart_constants.js:13]: True, though I thought it would be good to keep track of the ads that were targeted generally at the US as a whole
